policy-controller/src/main.rs

```diff
@@ -111,17 +121,70 @@ async fn main() -> Result<()> {
     // Run the gRPC server, serving results by looking up against the index handle.
     tokio::spawn(grpc(grpc_addr, cluster_networks, handle, drain_rx.clone()));
 
-    // Run the admission controller
     if let Some(bind_addr) = admission_addr {
-        let (listen_addr, serve) = warp::serve(admission::routes(client))
-            .tls()
-            .cert_path("/var/run/linkerd/tls/tls.crt")
-            .key_path("/var/run/linkerd/tls/tls.key")
-            .bind_with_graceful_shutdown(bind_addr, async move {
-                let _ = drain_rx.signaled().await;
-            });
+        let tcp = TcpListener::bind(&bind_addr).await?;
+        let listen_addr = tcp.local_addr()?;
+        // Prepare a long-running future stream to accept and serve clients.
+        let incoming_tls_stream = stream! {
+            loop {
+                let socket = match tcp.accept().await {
+                    Ok((socket, _)) => socket,
+                    Err(err) => {
+                        error!(%err, "Failed to accept connection");
+                        continue;
+                    }
+                };
+                let tls_cfg = {
+                    // Load public certificate.
+                    let certs = match load_certs("/var/run/linkerd/tls/tls.crt") {
+                        Ok(certs) => certs,
+                        Err(err) => {
+                            error!(%err, "Failed to load certificate");
+                            continue;
+                        }
+                    };
+                    // Load private key.
+                    let key = match load_private_key("/var/run/linkerd/tls/tls.key") {
+                        Ok(key) => key,
+                        Err(err) => {
+                            error!(%err, "Failed to load key");
+                            continue;
+                        }
+                    };
+                    // Do not use client certificate authentication.
+                    let mut cfg = match rustls::ServerConfig::builder()
+                        .with_safe_defaults()
+                        .with_no_client_auth()
+                        .with_single_cert(certs, key) {
+                        Ok(cfg) => cfg,
+                        Err(err) => {
+                            error!(%err, "Failed to configure TLS");
+                            continue;
+                        }
+                    };
+                    // Configure ALPN to accept HTTP/2, HTTP/1.1 in that order.
+                    cfg.alpn_protocols = vec![b"h2".to_vec(), b"http/1.1".to_vec()];
+                    sync::Arc::new(cfg)
+                };
+                let tls_acceptor = TlsAcceptor::from(tls_cfg);
+                let stream = tls_acceptor.accept(socket);
+                match stream.await {
+                    Ok(stream) => yield Ok::<_, Infallible>(stream),
+                    Err(e) => error!(%e, "TLS Error"),
```

Comment:
nit: the name `e` will be included in the formatted output as the field name for the error. we might want to call it
```suggestion
                    Err(err) => error!(%err, "TLS Error"),
```
or
```suggestion
                    Err(error) => error!(%error, "TLS Error"),
```
instead, so that the log line doesn't say
```
TLS Error e=<the actual error message>
```